File: backend/api/schemas/patents.py
"""Pydantic models for patent search, profile, and filter endpoints."""

from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class PatentCardModel(BaseModel):
    """Compact patent card for search results and selectable lists."""

    model_config = ConfigDict(extra="ignore")

    analysis_id: str
    patent_id: str
    source: str
    source_authority: str
    import_method: str
    title: str
    abstract_preview: str
    assignee: str | None = None
    country: str | None = None
    year: str | None = None
    keywords: list[str] = Field(default_factory=list)
    candidate_application_areas: list[str] = Field(default_factory=list)
    match_label: str = "Curated record"
    match_score: int = 0
    source_url: str | None = None


class PatentSearchResponse(BaseModel):
    """JSON shape returned by GET /api/patents and /api/patents/search."""

    model_config = ConfigDict(extra="ignore")

    query: str
    total_results: int
    returned_results: int
    filters: dict[str, str]
    patents: list[PatentCardModel] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    limit: int | None = None
    offset: int = 0


class RelatedPatentModel(BaseModel):
    """Related patent overlap signal exposed via API."""

    model_config = ConfigDict(extra="ignore")

    analysis_id: str
    patent_id: str
    source: str
    source_authority: str
    title: str
    assignee: str | None = None
    country: str | None = None
    year: str | None = None
    relationship_strength: str
    similarity_score: float
    overlap_signal: str
    overlap_score: float
    same_technology_group: bool
    source_technology_group_id: int | None = None
    target_technology_group_id: int | None = None
    source_technology_group: str
    target_technology_group: str
    shared_keywords: list[str] = Field(default_factory=list)
    candidate_application_areas: list[str] = Field(default_factory=list)
    explanation: str = ""


class PatentProfileResponse(BaseModel):
    """Full profile for a single patent."""

    model_config = ConfigDict(extra="ignore")

    analysis_id: str
    patent_id: str
    source: str
    source_authority: str
    import_method: str
    title: str
    abstract: str
    plain_language_summary: str
    assignee: str | None = None
    inventors: list[str] = Field(default_factory=list)
    publication_date: str | None = None
    filing_date: str | None = None
    year: str | None = None
    country: str | None = None
    language: str | None = None
    source_url: str | None = None
    keywords: list[str] = Field(default_factory=list)
    ipc_codes: list[str] = Field(default_factory=list)
    cpc_codes: list[str] = Field(default_factory=list)
    claims_preview: str | None = None
    top_terms: list[dict[str, Any]] = Field(default_factory=list)
    candidate_application_areas: list[dict[str, Any]] = Field(default_factory=list)
    related_patents: list[RelatedPatentModel] = Field(default_factory=list)
    metadata_rows: list[dict[str, str]] = Field(default_factory=list)
    advanced_metadata_rows: list[dict[str, str]] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class RelatedPatentsResponse(BaseModel):
    """Related patents response for /api/patents/{analysis_id}/related."""

    model_config = ConfigDict(extra="forbid")

    analysis_id: str
    related_patents: list[RelatedPatentModel] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class FilterOptionsResponse(BaseModel):
    """Available filter options for the patent search UI."""

    model_config = ConfigDict(extra="forbid")

    sources: list[str] = Field(default_factory=list)
    source_counts: dict[str, int] = Field(default_factory=dict)
    assignees: list[str] = Field(default_factory=list)
    top_assignees: dict[str, int] = Field(default_factory=dict)
    countries: list[str] = Field(default_factory=list)
    country_counts: dict[str, int] = Field(default_factory=dict)
    years: list[str] = Field(default_factory=list)
    publication_year_range: dict[str, int | None] = Field(default_factory=dict)
    filing_year_range: dict[str, int | None] = Field(default_factory=dict)
    top_keywords: dict[str, int] = Field(default_factory=dict)
    top_application_areas: dict[str, int] = Field(default_factory=dict)
    classifications: list[str] = Field(default_factory=list)
    top_classifications: dict[str, int] = Field(default_factory=dict)
    technology_groups: list[str] = Field(default_factory=list)
    candidate_application_areas: list[str] = Field(default_factory=list)

File: frontend/src/types/patents.ts
export interface PatentCard {
  analysis_id: string;
  patent_id: string;
  source: string;
  source_authority: string;
  import_method: string;
  title: string;
  abstract_preview: string;
  assignee: string | null;
  country: string | null;
  year: string | null;
  keywords: string[];
  candidate_application_areas: string[];
  match_label: string;
  match_score: number;
  source_url: string | null;
}

export interface PatentSearchResponse {
  query: string;
  total_results: number;
  returned_results: number;
  filters: Record<string, string>;
  patents: PatentCard[];
  warnings: string[];
  limit: number | null;
  offset: number;
}

export interface RelatedPatent {
  analysis_id: string;
  patent_id: string;
  source: string;
  source_authority: string;
  title: string;
  assignee: string | null;
  country: string | null;
  year: string | null;
  relationship_strength: string;
  similarity_score: number;
  overlap_signal: string;
  overlap_score: number;
  same_technology_group: boolean;
  source_technology_group_id: number | null;
  target_technology_group_id: number | null;
  source_technology_group: string;
  target_technology_group: string;
  shared_keywords: string[];
  candidate_application_areas: string[];
  explanation: string;
}

export interface ProfileTopTerm {
  term: string;
  importance: string;
  score: number;
}

export interface ProfileApplicationArea {
  area_name: string;
  score: number;
  evidence_level: string;
  matched_terms: string[];
  evidence_count: number;
  explanation: string;
}

export interface MetadataRow {
  Field: string;
  Value: string;
}

export interface PatentProfile {
  analysis_id: string;
  patent_id: string;
  source: string;
  source_authority: string;
  import_method: string;
  title: string;
  abstract: string;
  plain_language_summary: string;
  assignee: string | null;
  inventors: string[];
  publication_date: string | null;
  filing_date: string | null;
  year: string | null;
  country: string | null;
  language: string | null;
  source_url: string | null;
  keywords: string[];
  ipc_codes: string[];
  cpc_codes: string[];
  claims_preview: string | null;
  top_terms: ProfileTopTerm[];
  candidate_application_areas: ProfileApplicationArea[];
  related_patents: RelatedPatent[];
  metadata_rows: MetadataRow[];
  advanced_metadata_rows: MetadataRow[];
  warnings: string[];
}

export interface RelatedPatentsResponse {
  analysis_id: string;
  related_patents: RelatedPatent[];
  warnings: string[];
}

export interface FilterOptionsResponse {
  sources: string[];
  source_counts: Record<string, number>;
  assignees: string[];
  top_assignees: Record<string, number>;
  countries: string[];
  country_counts: Record<string, number>;
  years: string[];
  publication_year_range: {
    min: number | null;
    max: number | null;
  };
  filing_year_range: {
    min: number | null;
    max: number | null;
  };
  top_keywords: Record<string, number>;
  top_application_areas: Record<string, number>;
  classifications: string[];
  top_classifications: Record<string, number>;
  technology_groups: string[];
  candidate_application_areas: string[];
}

export const ALL_FILTER_VALUE = "All";

export interface PatentSearchParams {
  q?: string;
  source?: string;
  assignee?: string;
  country?: string;
  year?: string;
  limit?: number;
  offset?: number;
}

export interface PatentListParams {
  limit?: number;
  offset?: number;
}

export function isFilterActive(params: PatentSearchParams): boolean {
  const q = (params.q ?? "").trim();
  if (q.length > 0) return true;
  const filterKeys: Array<keyof PatentSearchParams> = [
    "source",
    "assignee",
    "country",
    "year",
  ];
  return filterKeys.some((key) => {
    const value = params[key];
    return typeof value === "string" && value !== "" && value !== ALL_FILTER_VALUE;
  });
}
